man/vde_plug.1
.TH VDE_PLUG 1 "August 23, 2016" "Virtual Distributed Ethernet"
.SH NAME
vde_plug \- Virtual Distributed Ethernet plug (two plugs creates a vde cable)
.SH SYNOPSIS
.B vde_plug
[
.I OPTIONS
]
[
.I VNL
]
.br
.B vde_plug
[
.I OPTIONS
]
.I VNL
.I VNL
.br
.B vde_plug
[
.I OPTIONS
]
.B =
.I command
[
.I args
]
.br
.B vde_plug
[
.I OPTIONS
]
.I VNL
.B =
.I command
[
.I args
]
.br
.SH DESCRIPTION
A
\fBvde_plug\fR
is a plug to be connected into a VDE network.

VDE network sockets are named using the Virtual Network Locator (VNL)
syntax, i.e.  \fImodule\fB://\fIspecific_address\fR.
e.g. \fBvde:///home/user/myswitch\fR or \fBvxvde://239.0.0.1\fR.
The default module is \fBlibvdeplug_vde(1)\fR (connection to a \fBvde_switch(1)\fR)
and can be omitted. e.g. \fB/home/user/myswitch\fR means
\fBvde:///home/user/myswitch\fR

\fBvde_plug\fR can have zero, one or two VNL arguments.
In case of zero or one argument the network communication is converted in
a byte stream. Packets from the VDE network are sent to stdout and bytes
from stdin are converted into packets and injected in the VDE network.
When a VNL is omitted or it is an empty argument
(\fB''\fR) \fBvde_plug\fR tries to connect to a
default network (defined by the user in \fB$HOME/.vde2/default.switch\fR
otherwise \fBvde:///run/vde.ctl\fR or \fBvde:///tmp/vde.ctl\fR).

This tool has been designed to be used together with
.B dpipe (1)
to interconnect a second vde_plug to another switch, working as a
virtual ethernet crossed cable between the two switches.

The command
.RS
.br
.B dpipe vde_plug = vde_plug vde:///tmp/vde2.ctl
.RE
connects two local switches: the former is using the standard control socket /tmp/vde.ctl
(or /var/run/vde.ctl) while the latter is using /tmp/vde2.ctl.

\fBvde_plug\fR creates a virtual cable between to VDE networks
when two VNL arguments are present in the command line.
The previous command is equivalent to:
.RS
.br
.B vde_plug /tmp/vde.ctl /tmp/vde2.ctl
.RE

The following example connects a vxvde network to a tap interfave
.RS
.br
.B vde_plug vxvde://239.1.2.3 tap://mytap
.RE

The command
.RS
.br
.B dpipe vde_plug = ssh remote.machine.org vde_plug
.RE
connects two remote switches.
If for example the two vde_switches run as daemon and they are connected to tap interfaces
a level 2 encrypted tunnel is established.

While \fBdpipe\fR supports more complex vitrual network structures (e.g. including \fBwirefilter(1)\fR to
emulate network conditions) \fBvde_plug\fR has been designed to provide simple syntax options
to implement the most common virtual network scenarios.
It is possible to use an equal sign (\fB=\fR) followed by a command and its optional
arguments in place of the second VNL.
In this case the packets from/to the network are converted into a bidirectional byte stream
provided as stdin and stdout to the command.
The previous command have the same effect of:
.RS
.br
.B vde_plug = ssh remote.machine.org vde_plug
.RE
(the first VNL is omitted).

This example:
.RS
.br
.B vde_plug vxvde://239.1.2.3 = ssh vde_plug tap://remotetap
.RE
connects a vxvde network to a tap interface of a remote host.

vde_plug can also be established as a login shell for users.
The following command works as in the previous example
.RS
.br
.B vde_plug = ssh vdeuser@remote.machine.org vde_plug
.RE
where vdeuser is the user with vde_plug as standard shell.
All the remote accesses are logged by syslog at the beginning and at the
end of each session and the IP address used are logged if \fBvdeuser\fR belongs
to the \fBvdeplug_iplog\fR group.
Attempts to login without the command vde_plug at the end or to
run on the remote host other commands but vde_plug
are blocked and the violation is logged by syslog.

.SH OPTIONS
.TP
\fB\-d
.TQ
\fB\-\-daemon
run as a daemon.
.TP
\fB\-p\fR \fIPIDFILE
.TQ
\fB\-\-pidfile\fR \fIPIDFILE
write the process id to the file \fIPIDFILE\fR
.TP
\fB\-l
.TQ
\fB\-\-log
log START/STOP of vde_plug on syslog
.TP
\fB\-L
.TQ
\fB\-\-iplog
log START/STOP of vde_plug and the IP addresses of hosts seen on the stream
(or on the second VNL) on syslog.
.TP
\fB\-\-port "\fIportnum\fP"
Obsolete option, use the syntax of the libvdeplug_vde module instead,
add the port number in square brackets suffix (e.g. \fBvde://tmp/myswitch[10]\fR).
It is possible to decide which port of the switch to use.
When this option is not specified the switch assigns the first
available unused port (if any).
It is possible to connect several cables in the same switch port:
in this way all this cables work concurrently.
It means that packet can result as duplicate but no ARP table loops are
generated.
Is useful when vde is used for mobility. Several physical interfaces can
be used at a time during handoffs to prevent hichups in connectivity.
log START/STOP of vde_plug on syslog
.TP
\fB\-g \fIgroup
.TQ
\fB\-\-group \fIgroup
group ownership of the communication socket. For security when more
want to share a switch it is better to use a unix group to own the comm sockets
so that the network traffic cannot be sniffed.
.TP
\fB\-m \fIoctal-mode\fR
.TQ
\fB\-\-mod \fIoctal-mode\fR
octal chmod like permissions for the comm sockets
.TP
\fB\-G \-M
.TQ
\fB\-\-port2 \-\-group2 \-\-mod2
These options, valid in the syntax with two VNLs, have the the same meaning of \-p \-g \-m.
The uppercase options or the long options with a trailing 2 refer to the second VNL.
.TP
\fB\-D\fR \fIDESCR
.TQ
\fB\-\-descr\fR \fIDESCR
set the description of this connection to \fIDESCR\fR (e.g. port/print command on a
vde_switch shows this description).


.SH NOTICE
Virtual Distributed Ethernet is not related in any way with
www.vde.com ("Verband der Elektrotechnik, Elektronik und Informationstechnik"
i.e. the German "Association for Electrical, Electronic & Information
Technologies").

.SH SEE ALSO
\fBvde_switch\fP(1),
\fBvdeq\fP(1),
\fBdpipe\fP(1),
\fBlibvdeplug_p2p.1\fR,
\fBlibvdeplug_udp.1\fR,
\fBlibvdeplug_vxlan.1\fR,
\fBlibvdeplug_tap.1\fR,
\fBlibvdeplug_vde.1\fR,
\fBlibvdeplug_vxvde.1\fR.
.br
.SH AUTHOR
VDE is a project by Renzo Davoli <renzo@cs.unibo.it>


debian/changelog
vdeplug4 (0.1-1) unstable; urgency=medium

  * Initial release (Closes: #0000)

 -- Mattia Biondi <mattia.biondi3@studio.unibo.it>  Sat, 14 Sep 2019 10:14:53 +0200


srcvdeplug4/vde_plug.c
/* 
 * Copyright (C) 2002-2016  Renzo Davoli, University of Bologna
 * Modified by Ludovico Gardenghi 2005
 * 
 * vde_plug: connect vde network together
 *
 * VDE is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; If not, see <http://www.gnu.org/licenses/>. 
 *
 */

#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <errno.h>
#include <unistd.h>
#include <stdint.h>
#include <getopt.h>
#include <string.h>
#include <sys/ioctl.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <sys/uio.h>
#include <sys/utsname.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <pwd.h>
#include <grp.h>
#include <stdarg.h>
#include <poll.h>

#include <libvdeplug.h>
#include <vde_plug_log.h>
#include <vde_plug_iplog.h>
#include <openclosepidfile.h>
#include <selfsighandler.h>

#define VDE_IP_LOG_GROUP "vdeplug_iplog"
#define DEFAULT_DESCRIPTION "vdeplug:"

int vde_ip_log;

VDECONN *conn;
VDECONN *conn2;
VDESTREAM *vdestream;

#define ETH_ALEN 6
#define ETH_HDRLEN (ETH_ALEN+ETH_ALEN+2)

struct utsname me;
#define myname me.nodename

void vdeplug_err(void *opaque, int type, char *format,...)
{
	va_list args;

	if (isatty(STDERR_FILENO)) {
		fprintf(stderr, "%s: Packet length error ",myname);
		va_start(args, format);
		vfprintf(stderr, format, args);
		va_end(args);
		fprintf(stderr,"\n");
	}
}

ssize_t vdeplug_recv(void *opaque, void *buf, size_t count)
{
	VDECONN *conn=opaque;
	if (__builtin_expect(vde_ip_log != 0, 0))
		vde_ip_check(buf,count);
	return vde_send(conn,(char *)buf,count,0);
}


int checkuidgroup(uid_t uid, const char *group) {
	struct passwd *pw=getpwuid(uid);
	int ngroups=0;
	if (pw == NULL) return -1;
	if (getgrouplist(pw->pw_name, pw->pw_gid, NULL, &ngroups) < 0) {
		gid_t gids[ngroups];
		if (getgrouplist(pw->pw_name, pw->pw_gid, gids, &ngroups) == ngroups) {
			struct group *grp;
			int i;
			while ((grp=getgrent()) != NULL) {
				for (i=0; i<ngroups; i++) {
					if (grp->gr_gid == gids[i] && strcmp(grp->gr_name,group) == 0) {
						endgrent();
						return 1;
					}
				}
			}
			endgrent();
			return 0;
		}
	}
	return -1;
}

static void cleanup(void)
{
  vde_close(conn);
	if (conn2 != NULL) vde_close(conn2);
	if (vdestream != NULL) vdestream_close(vdestream);
	openclosepidfile(NULL);
}

unsigned char bufin[VDE_ETHBUFSIZE];

int plug2stream(void) {
	register ssize_t nx;
	struct pollfd pollv[] = {
		{STDIN_FILENO, POLLIN|POLLHUP},
		{vde_datafd(conn), POLLIN|POLLHUP},
		{vde_ctlfd(conn), POLLIN|POLLHUP},
	};

	vdestream = vdestream_open(conn,STDOUT_FILENO,vdeplug_recv,vdeplug_err);

	for(;;) {
		poll(pollv,3,-1);
		if (__builtin_expect(((pollv[0].revents | pollv[1].revents | pollv[2].revents) & POLLHUP ||
						pollv[2].revents & POLLIN),0))
			break;
		if (pollv[0].revents & POLLIN) {
			nx = read(STDIN_FILENO,bufin,sizeof(bufin));
			/* if POLLIN but not data it means that the stream has been
			 * closed at the other end */
			/*fprintf(stderr,"%s: RECV %d %x %x \n",myname,nx,bufin[0],bufin[1]);*/
			if (nx == 0)
				break;
			vdestream_recv(vdestream, bufin, nx);
		}
		if (pollv[1].revents & POLLIN) {
			nx = vde_recv(conn,bufin,VDE_ETHBUFSIZE,0);
			if (__builtin_expect((nx >= ETH_HDRLEN),1))
			{
				vdestream_send(vdestream, bufin, nx);
				/*fprintf(stderr,"%s: SENT %d %x %x \n",myname,nx,bufin[0],bufin[1]);*/
			} else if (nx<0)
				perror("vde_plug: recvfrom ");
		}
	}
	return(0);
}

int plug2cmd(char **argv) {
	int p2c[2],c2p[2];
	pid_t pid;
	if (pipe(p2c) < 0 || pipe(c2p) < 0) {
		perror("pipe open");
		return(1);
	}

	pid=fork();
	if (pid < 0) {
		perror("fork");
		exit(1);
	} else if (pid == 0) {
		dup2(p2c[0], STDIN_FILENO);
		dup2(c2p[1], STDOUT_FILENO);
		close(p2c[0]); close(p2c[1]);
		close(c2p[0]); close(c2p[1]);
		execvp(*argv, argv);
		exit(1);
	}
	dup2(c2p[0], STDIN_FILENO);
	dup2(p2c[1], STDOUT_FILENO);
	close(p2c[0]); close(p2c[1]);
	close(c2p[0]); close(c2p[1]);

	return plug2stream();
}

int plug2plug(void)
{
	register ssize_t nx;
	struct pollfd pollv[] = {
		{vde_datafd(conn), POLLIN|POLLHUP},
		{vde_datafd(conn2), POLLIN|POLLHUP},
		{vde_ctlfd(conn), POLLIN|POLLHUP},
		{vde_ctlfd(conn2), POLLIN|POLLHUP}
	};

	for(;;) {
		poll(pollv,4,-1);
		if ((pollv[0].revents | pollv[1].revents | pollv[2].revents | pollv[2].revents) & POLLHUP ||
				(pollv[2].revents | pollv[3].revents) & POLLIN)
			break;
		if (pollv[0].revents & POLLIN) {
			nx = vde_recv(conn, bufin, VDE_ETHBUFSIZE,0);
			if (__builtin_expect((nx >= ETH_HDRLEN),1)) {
				vde_send(conn2, bufin, nx, 0);
				/*fprintf(stderr,"0->1 %d ",nx);*/
			} else if (nx<0)
				break;
		}
		if (pollv[1].revents & POLLIN) {
			nx = vde_recv(conn2,bufin,VDE_ETHBUFSIZE,0);
			if (__builtin_expect((nx >= ETH_HDRLEN),1)) {
				vde_send(conn, bufin, nx, 0);
				/*fprintf(stderr,"1->0 %d ",nx);*/
			} else if (nx<0)
				break;
		}
	}
	return(0);
}


static void netusage_and_exit() {
	vdeplug_openlog("FAILED");
	fprintf (stderr,"This is a Virtual Distributed Ethernet (vde) tunnel broker. \n"
			"This is not a login shell, only vde_plug can be executed\n");
	exit(-1);
}

static void usage_and_exit(char *progname) {
	fprintf (stderr,"Usage:\n"
			"  %s [ OPTIONS ] \n"
			"  %s [ OPTIONS ] vde_plug_url\n"
			"  %s [ OPTIONS ] vde_plug_url vde_plug_url\n"
			"  %s [ OPTIONS ] = command [args ...]\n"
			"  %s [ OPTIONS ] vde_plug_url = command [args ...]\n"
			"  an omitted or empty '' vde_plug_url refer to the default vde service\n"
			"  i.e. what defined in \"~/.vde2/default.switch\" or\n"
			"       a standard switch defined by libvdeplug_vde.so (e.g. /var/run/vde.ctl)\n"
			"Options:\n"
			"  -d | --daemon:            daemonize the program\n"
			"  -p PIDFILE | --pidfile PIDFILE: \n"
			"                            write pid of daemon to PIDFILE\n"
			"  -l | --log:               log START/STOP of vde_plug on syslog\n"
			"  -L | --iplog:             -l plus log the IP addresses used\n"
			"  -h | --help:              show this short summary\n"
			"  -g GROUP | -group GROUP:  set the group ownership (for vde://...)\n"
			"  -G GROUP | -group2 GROUP: like -g, for the second plug\n"
			"  -m MODE | -mod MODE:      comm socket protection mode (see chmod) (for vde://...)\n"
			"  -M MODE | -mod2 MODE:     like -m, for the second plug\n"
			"  --port PORT1, --port2 PORT2:\n"
			"                            obsolete options, set the vde switch port,\n"
			"                            use [port] suffix in the vde_plug_url instead\n"
			"  -D DESCR | --descr DESCR  set the description of this connection to DESCR\n"
			"                            (the default value is \"" DEFAULT_DESCRIPTION "\"\n"
			"\n",progname,progname,progname,progname,progname);
	exit(-1);
}

char short_options[] = "c:hp:dm:M:g:G:lLD:";
struct option long_options[] = {
	{"pidfile",  required_argument, 0, 'p' },
	{"daemon",   no_argument,       0, 'd' },
	{"port",     required_argument, 0, 0x100 + 'p'},
	{"port2",    required_argument, 0, 0x100 + 'P'},
	{"mod",      required_argument, 0, 'm'},
	{"mod2",     required_argument, 0, 'M'},
	{"group",    required_argument, 0, 'g'},
	{"group2",   required_argument, 0, 'G'},
	{"descr",    required_argument, 0, 'D' },
	{"log",      no_argument,       0, 'l' },
	{"iplog",    no_argument,       0, 'L' },
	{"help",     no_argument,       0, 'h' },
	{0, 0, 0, 0}
};

#include <syslog.h>
int main(int argc, char *argv[])
{
	char *progname = basename(argv[0]);
	static char *vde_url = NULL;
	static char *vde_url2 = NULL;
	static char **cmdargv = NULL;
	struct vde_open_args open_args = {.port = 0,.group = NULL,.mode = 0700};
	struct vde_open_args open_args2 = {.port = 0,.group = NULL,.mode = 0700};
	int daemonize = 0;
	char *pidfile = NULL;
	char *description = DEFAULT_DESCRIPTION;

	uname(&me);

	if (argv[0][0] == '-')
		netusage_and_exit(); //implies exit
	/* option parsing */
	while (1) {
		int option_index = 0;
		int c;

		c = getopt_long(argc, argv, short_options, long_options, &option_index);
		if (c == -1) break;

		switch (c) {
			case 'c':
				if (strcmp(optarg,"vde_plug") == 0) {
					vdeplug_openlog(NULL);
					atexit(vdeplug_closelog);
					if (checkuidgroup(getuid(), VDE_IP_LOG_GROUP) == 1)
						vde_ip_log = 1;
				}
				else
					netusage_and_exit(); //implies exit
				break;

			case 0x100 + 'p':
				open_args.port = atoi(optarg);
				if (open_args.port <= 0)
					usage_and_exit(progname); //implies exit
				break;

			case 0x100 + 'P':
				open_args2.port = atoi(optarg);
				if (open_args2.port <= 0)
					usage_and_exit(progname); //implies exit
				break;

			case 'h':
				usage_and_exit(progname); //implies exit
				break;

			case 'm': 
				sscanf(optarg,"%o",(unsigned int *)&(open_args.mode));
				break;

			case 'M': 
				sscanf(optarg,"%o",(unsigned int *)&(open_args2.mode));
				break;

			case 'g':
				open_args.group = optarg;
				break;

			case 'G':
				open_args2.group = optarg;
				break;

			case 'L':
				vde_ip_log = 1;

			case 'l':
				vdeplug_openlog(NULL);
				atexit(vdeplug_closelog);
				break;

			case 'p' : 
				pidfile = optarg;
				break;

			case 'd' : 
				daemonize = 1;
				break;

			case 'D' : 
				if (*optarg != 0)
					description = optarg;
				break;

			default:
				usage_and_exit(progname); //implies exit
		}
	}

	argv += optind;
	argc -= optind;

	switch (argc) {
		case 0: break;
		case 1: vde_url = *argv;
						break;
		case 2: if (strcmp(argv[0],"=") == 0)
							cmdargv = argv+1;
						else {
							vde_url = argv[0];
							vde_url2 = argv[1];
						}
						break;
		default: if (strcmp(argv[0],"=") == 0)
							 cmdargv = argv+1;
						 else if (strcmp(argv[1],"=") == 0) {
							 vde_url = argv[0];
							 cmdargv = argv+2;
						 } else
							 usage_and_exit(progname); //implies exit
						 break;
	}

	if (daemonize != 0) {
		if (daemon(0,0) < 0) {
			fprintf(stderr,"%s daemonize: %s\n", progname, strerror(errno));
			exit(1);
		}
	}

	if (pidfile != NULL)
		openclosepidfile(pidfile);

	atexit(cleanup);
	setsighandlers(cleanup);

	conn = vde_open(vde_url, description, &open_args);
	if (conn == NULL) {
		fprintf(stderr,"vde_open %s: %s\n",vde_url && *vde_url ? vde_url : "default switch", strerror(errno));
		exit(1);
	}
	if (vde_url2 != NULL) {
		conn2 = vde_open(vde_url2, description, &open_args2);
		if (conn2 == NULL) {
			fprintf(stderr,"vde_open %s: %s\n",vde_url2 && *vde_url2 ? vde_url2 : "default switch", strerror(errno));
			exit(1);
		}
		return plug2plug();
	} else if (cmdargv != NULL)
		return plug2cmd(cmdargv);
	else
		return plug2stream();
}
